pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapfillProcessor.java

```diff
@@ -0,0 +1,473 @@
+        } else {
+          gapfillRow[i] = getFillValue(i, dataSchema.getColumnName(i), key, resultColumnDataTypes[i]);
+        }
+      }
+
+      if (postGapfillFilterHandler == null || postGapfillFilterHandler.isMatch(gapfillRow)) {
+        if (bucketedResult.size() >= _limitForGapfilledResult) {
+          break;
+        } else {
+          bucketedResult.add(gapfillRow);
+        }
+      }
+    }
+    if (_limitForGapfilledResult > _groupByKeys.size()) {
+      _limitForGapfilledResult -= _groupByKeys.size();
+    } else {
+      _limitForGapfilledResult = 0;
+    }
+  }
+
+  private List<Object[]> aggregateGapfilledData(Object timeCol, List<Object[]> bucketedRows, DataSchema dataSchema) {
+    List<ExpressionContext> groupbyExpressions = _queryContext.getGroupByExpressions();
+    Preconditions.checkArgument(groupbyExpressions != null, "No GroupBy Clause.");
+    Map<String, Integer> indexes = new HashMap<>();
+    for (int i = 0; i < dataSchema.getColumnNames().length; i++) {
+      indexes.put(dataSchema.getColumnName(i), i);
+    }
+
+    for (Object [] bucketedRow : bucketedRows) {
+      bucketedRow[_timeBucketColumnIndex] = timeCol;
+    }
+
+    Map<List<Object>, Integer> groupKeyIndexes = new HashMap<>();
+    int[] groupKeyArray = new int[bucketedRows.size()];
+    List<Object[]> aggregatedResult = new ArrayList<>();
+    for (int i = 0; i < bucketedRows.size(); i++) {
+      Object[] bucketedRow = bucketedRows.get(i);
+      List<Object> groupKey = new ArrayList<>(groupbyExpressions.size());
+      for (ExpressionContext groupbyExpression : groupbyExpressions) {
+        int columnIndex = indexes.get(groupbyExpression.toString());
+        groupKey.add(bucketedRow[columnIndex]);
+      }
+      if (groupKeyIndexes.containsKey(groupKey)) {
+        groupKeyArray[i] = groupKeyIndexes.get(groupKey);
+      } else {
+        // create the new groupBy Result row and fill the group by key
+        groupKeyArray[i] = groupKeyIndexes.size();
+        groupKeyIndexes.put(groupKey, groupKeyIndexes.size());
+        Object[] row = new Object[_queryContext.getSelectExpressions().size()];
+        for (int j = 0; j < _queryContext.getSelectExpressions().size(); j++) {
+          ExpressionContext expressionContext = _queryContext.getSelectExpressions().get(j);
+          if (expressionContext.getType() != ExpressionContext.Type.FUNCTION) {
+            row[j] = bucketedRow[indexes.get(expressionContext.toString())];
+          }
+        }
+        aggregatedResult.add(row);
+      }
+    }
+
+    Map<ExpressionContext, BlockValSet> blockValSetMap = new HashMap<>();
+    for (int i = 1; i < dataSchema.getColumnNames().length; i++) {
+      blockValSetMap.put(ExpressionContext.forIdentifier(dataSchema.getColumnName(i)),
+          new RowBasedBlockValSet(dataSchema.getColumnDataType(i), bucketedRows, i));
+    }
+
+    for (int i = 0; i < _queryContext.getSelectExpressions().size(); i++) {
```

Comment:
You should be able to use the `_queryContext.getAggregationFunctions()`

Reply:
I need the selection index of aggregationFunction.